### ISSUE TITLE
Referring to a required attribute from BUILD, but not explicitly providing a value for it, initializes it in some cases

### ISSUE THREAD

**Issue by Kaiepi** [rakudo/rakudo#4624]
## The Problem

Consider:

```raku
class Box {
    has Any:D $.value is required;
    submethod BUILD(::?CLASS:D:) { $!value }
}.new
```

## Expected Behavior

This should throw an error either pertaining to `$!value`'s lack of value or its indefiniteness.

## Actual Behavior

This silently succeeds. Weirdly, giving `BUILD` `--> Nil` produces a missing attribute error (sunk otherwise?).

## Steps to Reproduce

Save the golf and run it.

## Environment
* Operating system:
```
OpenBSD bastille.kennel.qt 7.0 GENERIC.MP#47 amd64
```
* Compiler version (`perl6 -v` or `raku -v`):
```
Welcome to Rakudo™ v2021.10-52-ge98e17da7.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2021.10-28-g6ff815576.
```

**Comment by jnthn** [comment]
The way things currently work is that attributes all start out nulled (MoarVM uses a real `NULL` to model this situation), and are vivified on first access. While the presumption is that one will assign to the attribute, as you've discovered one could actually just mention it and it would count as having been touched and thus initialized. Not entirely ideal, but apparently good enough that we've pretty much gotten away with it for years. :-)

This mechanism has bothered me for a while for other reasons, however. For one, it increases the cost of every single attribute access (both in terms of time and code size), because we have to do the "is it NULL" check; those are not *that* costly, the extra branches take up branch prediction slots in the produced JIT output too. But it gets even worse when one comes to escape analysis. I set about trying to handle it there, and one has to try and track vivification status, figure out if it can be merged at PHIs, etc. (This work never yet made it into `master`, and today we just don't try and do EA in any such situation, which is hugely limiting.)

The problem has always been how to replace it, and for a long while I was pretty stuck for ideas...until recently, while working on `new-disp` and re-working the dispatchers handling assignment. A container descriptor hangs off a `Scalar` and says what it means to do an assignment into it (in terms of type checking, but it is also used as part of auto-vivification or array and hash elements). I realized we could use this as an alternative solution to attribute initialization too, by:

1. For attributes that need a `required` check or to run a default, create a `Scalar` (or `Array`, or `Hash`) with a descriptor type that indicates it is uninitialized (and probably points to the real descriptor).
2. Upon an assignment, update the descriptor to point to the "standard" one
3. The replacement for `attrinited` is thus to check if we a) have the expected container type (something else wasn't bound in its place) and b) that it still has the uninitialized descriptor. If so, we consider it uninitialized.

This would fix the problem reported in this issue, rid us of the need for the auto-vivification for semantics (removing it would be a separate step), and also would specialize rather nicely in many cases in combination with escape analysis. For example, if the object being constructed undergoes scalar replacement (perhaps only with delayed materialization - that is to say, at the end of the constructor we really construct the object), then we'd be able to see that the `Scalar` was created with the "uninitialized" descriptor, that this was never changed, and then constant-fold away the check of whether we needed to run the default. It'd not only avoid complexities in the EA, but also work naturally with it to get better results.
